bot.py

```diff
@@ -39,6 +37,145 @@ async def hello(ctx):
     await ctx.send("hi")
 
 
+@bot.command()
+async def count_members(ctx, *, role_names_str: str):
+    """Reads the message from the chat and returns the number of members
+    that meet the conditions.
+
+    Reads the message if the message starts with the "command_prefix"
+    (that was set in when initializing commands.Bot object) and the name
+    in the title of the method. Returns str in "ctx.send()".
+    """
+    not_roles = []
+    if " not " in role_names_str:
+        role_names = role_names_str.split(" not ")
+        role_names = [role.strip() for role in role_names if role]
+        not_role_names = [role for role in role_names[1:] if " not " not in role]
+        not_roles = [
+            discord.utils.get(ctx.guild.roles, name=not_role_name)
+            for not_role_name in not_role_names
+        ]
+        not_roles = [role for role in not_roles if role is not None]
+        role_names_str = role_names[0]
+    else:
+        not_roles = []
+
+    role_names = (
+        role_names_str.split(" and ")
+        if " and " in role_names_str
+        else role_names_str.split(" or ")
+    )
+    operator = "and" if " and " in role_names_str else "or"
+
+    roles = [
+        discord.utils.get(ctx.guild.roles, name=role_name) for role_name in role_names
+    ]
+    roles = [role for role in roles if role is not None]
+    if not roles:
+        await ctx.send(
+            f"Could not find any of the roles with names `{role_names_str}`."
+        )
+        return
+
+    members = set()
+    if operator == "and":
+        role_names = " and ".join(role.name for role in roles)
+        for role in roles:
+            if not members:
+                members.update(role.members)
+            else:
+                members &= set(role.members)
+    else:
+        role_names = " or ".join(role.name for role in roles)
+        for role in roles:
+            members.update(role.members)
+
+    for not_role in not_roles:
+        members -= set(not_role.members)
```

Comment:
Dobrze by było, gdybyś wydzielił kod generujący listę członków do oddzielnej funkcji.
Wtedy nie będziesz miał powtarzającego się kodu w `count_members` i `list_members`

Reply:
Poza tym funkcje będą dużo krótsze i czytelniejsze.

Reply:
Ten komentarz jest wciąż aktualny.    
Tworzenie zbioru pasujących użytkowników `members` nadal powinno być wydzielone

Reply:
Proponowałbym to podzielić na 2-3 funkcje max.

> "Tworzenie zbioru pasujących użytkowników members nadal powinno być wydzielone"
>
> Rozwiń proszę. Mówisz o podfunkcji? Dlaczego akurat specyficznie to z całej funkcji musi być wydzielone?

Mówię o stworzeniu oddzielnej podfunkcji. Akurat to musi być wydzielone, ponieważ jest to fragment, który się powtarza w dwóch miejscach. Opisałem to w wątku. Jest to blok, kodu, który pojawia się dwa razy i jest skopiowany niemal 1 do jednego.